csharp/ql/src/semmle/code/csharp/exprs/Creation.qll

```diff
@@ -344,10 +344,20 @@ class ArrayInitializer extends Expr, @array_init_expr {
   override string getAPrimaryQlClass() { result = "ArrayInitializer" }
 }
 
+private newtype TArrayCreation = MkArrayCreation()
+
+abstract class ArrayCreationBase extends TArrayCreation {
+  abstract string toString();
+
+  abstract predicate hasKnownLength();
+
+  abstract int getKnownLength();
+}
+
 /**
  * An array creation, for example `new int[,] { {0, 1}, {2, 3}, {4, 5} }`.
  */
-class ArrayCreation extends Expr, @array_creation_expr {
+class ArrayCreation extends Expr, ArrayCreationBase, @array_creation_expr {
```

Comment:
@hvitved I thought about adding a new base class here, but this doesn't work. I'm getting this QL error:
`Class has invalid supertypes: Component TArrayCreation has universe TArrayCreation, but other components have universe entity`. How could I add a base class for `ArrayCreation` and `EmptyArrayCreationCall`? I would like to add two extra predicates to these classes through the base class.

Reply:
It doesn't work because IPA types and DB types (`TArrayCreation` and `@array_creation_expr`) are not compatible.  I think if you let `ArrayCreationBase` extend `Expr` instead, things should work.

Reply:
Thanks. I'll give that a try.